File: source/aws/py/lambda-docker-websocket/UniRpc/GroupAuthorizations.py
from typing import TypedDict, Union, List, Dict
from UniRpc.GroupAuthorizationPolicies import __PolicySets


class IGroupPolicy(TypedDict):
    Name: str
    Services: Dict[str, Union[str, List[str]]]


IGroupPolicySet = Dict[str, IGroupPolicy]

IPolicySets = Dict[str, IGroupPolicySet]


def GroupAuthorizations(policy: str, group: str, service: str, method: str) -> bool:
    if policy == '*' and group == '*':
        return True
    if policy not in __PolicySets.keys():
        return False
    groupPolicySet: IGroupPolicySet = __PolicySets[policy]
    if group not in groupPolicySet.keys():
        return False
    groupPolicy: IGroupPolicy = groupPolicySet[group]
    services = groupPolicy['Services']
    if service not in services.keys():
        return False
    servicePolicy: Union[str, List[str]] = services[service]
    if isinstance(servicePolicy, str) and servicePolicy == '*':
        return True
    if isinstance(servicePolicy, list) and servicePolicy.includes(method):
        return True
    return False


def GroupClausesAuthorize(groups: List[str], service: str, method: str) -> bool:
    if not isinstance(groups, list):
        return False
    for groupClause in groups:
        sections = groupClause.split('.')
        policy = sections[0]
        group = sections[1]
        if GroupAuthorizations(policy, group, service, method):
            return True
    return False

File: source/aws/rpc/transformers/typescript/http-express-service/GroupAuthorizations.ts
// import { __PolicySets } from './GroupAuthorizationPolicies';

// export interface IGroupPolicy {
//     Name: string;
//     Services: {[service: string]: string | string[]};
// }

// export interface IGroupPolicySet {
//     [group: string]: IGroupPolicy;
// }

// export interface IPolicySets {
//     [policy: string]: IGroupPolicySet;
// }

// export function GroupAuthorizations(policy: string, group: string, service: string, method: string): boolean {
//     if (policy == '*' && group == '*') return true;
//     if (!__PolicySets[policy]) return false;
//     let groupPolicySet: IGroupPolicySet = __PolicySets[policy];
//     if (!groupPolicySet[group]) return false;
//     let groupPolicy: IGroupPolicy = groupPolicySet[group];
//     if (!groupPolicy.Services || !groupPolicy.Services[service]) return false;
//     let servicePolicy: string | string[] = groupPolicy.Services[service];
//     if (typeof servicePolicy == 'string' && servicePolicy == '*') {
//         return true;
//     }
//     if (Array.isArray(servicePolicy) && servicePolicy.includes(method)) {
//         return true;
//     }
//     return false;
// }

// export function GroupClausesAuthorize(groups: string[], service: string, method: string) {
//     if (!Array.isArray(groups)) return false;
//     for (let groupClause of groups) {
//         let sections = groupClause.split('.');
//         let policy = sections[0], group = sections[1];
//         if (GroupAuthorizations(policy, group, service, method)) return true;
//     }
//     return false;
// }